runtime/memprof.c

```diff
@@ -248,100 +268,111 @@ static value capture_callstack_major()
   return res;
 }
 
-static value capture_callstack_minor()
+static value capture_callstack_minor(void)
 {
   value res;
   intnat wosize = caml_current_callstack_size(callstack_size);
-  CAMLassert(caml_memprof_suspended); /* => no samples in the call stack. */
+  CAMLassert(!caml_memprof_suspended);
+  caml_memprof_suspended = 1; /* => no samples in the call stack. */
   res = caml_alloc(wosize, 0);
+  caml_memprof_suspended = 0;
   caml_current_callstack_write(res);
   return res;
 }
 
+/**** Handling postponed sampled blocks. ****/
+/* When allocating in from C code, we cannot call the callback,
+   because the [caml_alloc_***] are guaranteed not to do so. These
+   functions make it possible to register a sampled block in a
+   todo-list so that the callback call is performed when possible. */
+/* Note: the shorter the delay is, the better, because the block is
+   linked to a root during the delay, so that the reachability
+   properties of the sampled block are artificially modified. */
+
 struct caml_memprof_postponed_block {
   value block;
   value callstack;
   int32_t occurrences;
+  enum ml_alloc_kind kind;
   struct caml_memprof_postponed_block* next;
-} *caml_memprof_postponed_head = NULL;
+} *caml_memprof_postponed_hd = NULL, *caml_memprof_postponed_tl = NULL;
 
-/* When allocating in the major heap, we cannot call the callback,
-   because [caml_alloc_shr] is guaranteed not to call the GC. Hence,
-   this function determines if the block needs to be sampled, and if
-   so, it registers the block in the todo-list so that the callback
-   call is performed when possible. */
-void caml_memprof_track_alloc_shr(value block)
-{
-  int32_t occurrences;
-  CAMLassert(Is_in_heap(block));
-  /* This test also makes sure memprof is initialized. */
-  if(lambda == 0 || caml_memprof_suspended)
-    return;
-  occurrences = mt_generate_poisson(Whsize_val(block));
-  if(occurrences > 0) {
-    struct caml_memprof_postponed_block* pb =
+static void free_postponed_block(struct caml_memprof_postponed_block *p) {
+  caml_remove_generational_global_root(&p->callstack);
+  caml_remove_generational_global_root(&p->block);
+  caml_stat_free(p);
+}
+
+static void purge_postponed_queue(void) {
+  while(caml_memprof_postponed_hd != NULL) {
+    struct caml_memprof_postponed_block * next =
+      caml_memprof_postponed_hd->next;
+    free_postponed_block(caml_memprof_postponed_hd);
+    caml_memprof_postponed_hd = next;
+  }
+  caml_memprof_postponed_tl = NULL;
+}
+
+static void register_postponed_callback(value block, uintnat occurrences,
```

Comment:
If, as I understand it, this function cannot call the GC, then it would be nice to indicate it in a comment. If that requirement is somehow gone, then is it still necessary to use `capture_callstack_major()` below?